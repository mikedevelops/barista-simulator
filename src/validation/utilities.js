import { Map } from 'immutable';

/**
 * Starting point for all validation
 * @type {number}
 */
export const VALIDATION_SCORE_BASE = 10;

/**
 * Create a new Validation score
 */
export const createValidationScore = () =>
    VALIDATION_SCORE_BASE;

/**
 * Subtract from validation score
 * @param validationScore
 * @param weight
 */
export const deductValidationScore = (validationScore, weight) =>
    validationScore - weight;

/**
 * Create Validation Result
 * @param score
 * @param messages
 */
export const createValidationResult = (score, messages) =>
    Map({ score, messages });
