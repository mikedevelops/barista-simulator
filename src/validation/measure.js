import { createValidationResult, createValidationScore, deductValidationScore } from './utilities'
import { List, Map } from 'immutable';

/**
 * Measure validation weight map
 * @type {*}
 */
export const measureValidationWeights = {
    AMOUNT: (expected, received) => {
        let diff = expected - received;

        if (diff < 0) {
            diff = diff * -1;
        }

        if (diff > 50) {
            return 3;
        }

        if (diff > 25) {
            return 2;
        }

        if (diff > 10) {
            return 1;
        }
    },
    SUBSTANCE: 3
};

/**
 * Validate a full measure
 *
 * @param recipeMeasure
 * @param orderMeasure
 * @param weightMap
 */
export const validateMeasure = (
    recipeMeasure,
    orderMeasure,
    weightMap = measureValidationWeights
) => {
    let validationMessages = List();
    let validationScore = createValidationScore();

    try {
        validateMeasureAmount(recipeMeasure, orderMeasure);
    } catch (err) {
        validationMessages = validationMessages.push(Map({
            message: err.message
        }));
        validationScore = deductValidationScore(
            validationScore,
            weightMap.AMOUNT(
                recipeMeasure.get('amount'),
                orderMeasure.get('amount')));
    }

    try {
        validateMeasureSubstance(recipeMeasure, orderMeasure);
    } catch (err) {
        validationMessages = validationMessages.push(Map({
            message: err.message
        }));
        validationScore = deductValidationScore(
            validationScore,
            weightMap.SUBSTANCE);
    }

    return createValidationResult(validationScore, validationMessages);
};

/**
 * Validate measure amount
 *
 * @param recipeMeasure
 * @param orderMeasure
 * @returns {boolean}
 */
export const validateMeasureAmount  = (recipeMeasure, orderMeasure) => {
    const recipeMeasureAmount = recipeMeasure.get('amount');
    const orderMeasureAmount = orderMeasure.get('amount');

    if (recipeMeasureAmount !== orderMeasureAmount) {
        throw new Error(`"${orderMeasureAmount}" is the wrong amount. ` +
            `Expected "${recipeMeasureAmount}"`);
    }

    return true;
}

/**
 * Validate Measure substance
 *
 * @param recipeMeasure
 * @param orderMeasure
 * @throws
 * @returns {boolean}
 */
export const validateMeasureSubstance  = (recipeMeasure, orderMeasure) => {
    const recipeMeasureSubstance = recipeMeasure.getIn(['substance', 'name']);
    const orderMeasureSubstance = orderMeasure.getIn(['substance', 'name']);

    if (recipeMeasureSubstance !== orderMeasureSubstance) {
        throw new Error(`"${orderMeasureSubstance}" is the wrong substance. ` +
            `Expected "${recipeMeasureSubstance}"`);
    }

    return true;
}
