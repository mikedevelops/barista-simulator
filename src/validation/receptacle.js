import { createValidationResult, createValidationScore, deductValidationScore } from './utilities'
import { List, Map } from 'immutable';

export const receptacleValidationWeights = {
    TYPE: () => -2
};

export const validateReceptacle = (
    recipeReceptacle,
    orderReceptacle,
    weightMap = receptacleValidationWeights
) => {
    let validationMessages = List();
    let validationScore = createValidationScore();

    try {
        validateReceptacleType(recipeReceptacle, orderReceptacle);
    } catch (err) {
        validationMessages = validationMessages.push(Map({
            message: err.message
        }));
        validationScore = deductValidationScore(
            validationScore,
            weightMap.TYPE());
    }

    return createValidationResult(validationScore, validationMessages);
}

/**
 * Validate Receptacle
 *
 * @param recipeReceptacle
 * @param orderReceptacle
 * @throws
 * @returns boolean
 */
export const validateReceptacleType = (recipeReceptacle, orderReceptacle) => {
    if (recipeReceptacle.get('name') !== orderReceptacle.get('name')) {
        throw new Error(`"${orderReceptacle.get('name')}" is the wrong receptacle. ` +
            `Expected a "${recipeReceptacle.get('name')}".`);
    }

    return true;
}
