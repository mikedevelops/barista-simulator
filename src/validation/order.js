import { createValidationScore } from './utilities';
import { validateReceptacle } from './receptacle'
import { validateMeasure } from './measure'

export const validateOrder = (expectedOrder, receivedReceptacle) => {
    // validate receptacle
    let score = validateReceptacle(expectedOrder.get('receptacle'), receivedReceptacle);

    // validate ingredients
    expectedOrder.getIn(['recipe', 'ingredients']).forEach(ingredient => {
        const receivedIngredient = receivedReceptacle.get('contents').find(content =>
                content.getIn(['substance', 'name']) === ingredient.getIn(['substance', 'name']));

        if (receivedIngredient !== undefined) {
            score = validateMeasure(ingredient, receivedIngredient);
        }
    });

    return score;
}
