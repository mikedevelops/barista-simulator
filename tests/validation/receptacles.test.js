import { validateReceptacleType } from '../../src/validation/receptacle'
import {
    createRegularGlassReceptacle, createRegularMugReceptacle,
    createSmallGlassReceptacle
} from '../../src/items/receptacles'

describe('receptacle validation', () => {
    describe('validateReceptacleType', () => {
        test('should throw an error if the receptacles do not match', () => {
            const recipeReceptacle = createRegularGlassReceptacle();
            const orderReceptacle = createRegularMugReceptacle();

            expect(() => validateReceptacleType(recipeReceptacle, orderReceptacle)).toThrowErrorMatchingSnapshot();
        });

        test('should return true for valid receptacle types', () => {
            const recipeReceptacle = createRegularGlassReceptacle();
            const orderReceptacle = createSmallGlassReceptacle();

            expect(validateReceptacleType(recipeReceptacle, orderReceptacle)).toBe(true);
        });
    });
});
