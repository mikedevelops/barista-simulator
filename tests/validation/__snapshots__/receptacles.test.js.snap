// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`receptacle validation validateReceptacleType should throw an error if the receptacles do not match 1`] = `"\\"RECEPTACLE_MUG\\" is the wrong receptacle. Expected a \\"RECEPTACLE_GLASS\\"."`;
